### ISSUE TITLE
How to random shuffle data correctly?

### ISSUE THREAD

**Issue by PistonY** [NVIDIA/DALI#1234]
Hello, I'm trying to use `nvidia.dali.ops.FileReader` load my data.
And it seems that you have so many attribute to control this like `initial_fill`, `random_shuffle`, `shuffle_after_epoch`, `stick_to_shard`, it's totally confuse me.
I only want DALI do the same thing with `torch.utils.data.DataLoader`. when setting `shuffle=True` which means shuffle data through the whole dataset at every epoch.
What should I do?

**Comment by awolant** [comment]
Hi, thanks for the question.
For the meaning of those attributes you can consult our docs for [`FileReader`](https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/supported_ops.html?highlight=filereader#nvidia.dali.ops.FileReader).

In your use case you need `random_shuffle=False` and `shuffle_after_epoch=True`. With this settings DALI will shuffle the whole dataset after every epoch.

For information on frameworks integration look into [PyTorch section](https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/plugins/pytorch_tutorials.html) of our docs.

**Comment by PistonY** [comment]
I've read all you docs. Thx for your reply.

If I use `random_shuffle=False` and `shuffle_after_epoch=True`, Will DALI shuffle data at first epoch?

**Comment by awolant** [comment]
Yes, it will. It will shuffle them first time during `FileReader` op construction.

**Comment by PistonY** [comment]
Thanks again.

I notice `FileReader` has `tensor_init_bytes`, I want to know if I train/val with 224x224x3 imgs, should I change it to `224x224x3x4`?

And `ImageDecoderRandomCrop` also has two attr `device_memory_padding` and `host_memory_padding`. Could you please explain how do you get the default values, when do you use it, and how to set them suitably(I've read the doc and still confuse)?

**Comment by awolant** [comment]
`FileReader` loads data from files to memory buffers. `tensor_init_bytes` parameter is used to control the initial size of those buffers. If sample from the current iteration  is too big for the buffer that was assigned to accommodate it, the buffer will be reallocated. Usually this happens in the initial phase of the training.
Please keep in mind, that reader deals with raw, not decoded data. The size depends on the format.

`device_memory_padding` and `host_memory_padding` do similar thing but for internal buffers used by `nvJPEG` to decode the images.

Unless you want to super fine tune your code for exact hardware, software stack and dataset it is ok to leave all of those parameters unset. Buffers will be resized accordingly during the training.

If you want to do it, you need to experimentally find best values.

**Comment by PistonY** [comment]
Ok, I’ll take care of that.Thx for your patient reply again.

**Comment by awolant** [comment]
@PistonY  No problem, happy to help.
I'm closing the issue for now. If you have any more comments or questions please do not hesitate to reopen or post a new one.

**Comment by PistonY** [comment]
@awolant I success use DALI to generate data.

Could you please tell me how to drop last batch if the last data can't fill in a full batch size.
Just like using `torch.utils.data.DataLoader` and setting `drop_last=True`.

**Comment by PistonY** [comment]
I've knew that.
DALI use size to control how many data can be generated and set `fill_last_batch=False` could avoid fill last batch.
